File: bank-service/src/main/java/rs/edu/raf/listener/RabbitMQListener.java
package rs.edu.raf.listener;

import com.google.gson.Gson;
import org.springframework.amqp.rabbit.annotation.RabbitListener;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import rs.edu.raf.dto.ListingOrderQueueDTO;
import rs.edu.raf.dto.PaymentQueueDTO;
import rs.edu.raf.service.BankAccountService;

@Component
public class RabbitMQListener {
    @Autowired
    private Gson gson;

    @Autowired
    private BankAccountService bankAccountService;

    @RabbitListener(queues = "${transactions.required.queue}")
    public void transaction(String message) {
        PaymentQueueDTO paymentQueueDTO = gson.fromJson(message, PaymentQueueDTO.class);
        bankAccountService.processPayment(paymentQueueDTO);
    }

    @RabbitListener(queues = "${update.balance.queue}")
    public void listingOrderTransaction(String message) {
        ListingOrderQueueDTO listingOrderQueueDTO = gson.fromJson(message, ListingOrderQueueDTO.class);

        bankAccountService.processListingOrder(listingOrderQueueDTO.buyerId(), listingOrderQueueDTO.sellerId(),
                listingOrderQueueDTO.buyPrice(), listingOrderQueueDTO.sellPrice());
    }
}

File: bank-service-kotlin/src/main/kotlin/org/example/bankservicekotlin/listener/RabbitMQListener.kt
package org.example.bankservicekotlin.listener

import com.google.gson.Gson
import org.example.bankservicekotlin.dto.ListingOrderQueueDTO
import org.example.bankservicekotlin.dto.PaymentQueueDTO
import org.example.bankservicekotlin.service.BankAccountService
import org.springframework.amqp.rabbit.annotation.RabbitListener
import org.springframework.stereotype.Component

@Component
class RabbitMQListener(
    val gson: Gson,
    val bankAccountService: BankAccountService) {
    @RabbitListener(queues = ["\${transactions.required.queue}"])
    fun transaction(message: String?) {
        val paymentQueueDTO: PaymentQueueDTO = gson.fromJson(message, PaymentQueueDTO::class.java)
        bankAccountService.processPayment(paymentQueueDTO)
    }

    @RabbitListener(queues = ["\${update.balance.queue}"])
    fun listingOrderTransaction(message: String?) {
        val listingOrderQueueDTO: ListingOrderQueueDTO = gson.fromJson(message, ListingOrderQueueDTO::class.java)

        bankAccountService.processListingOrder(
            listingOrderQueueDTO.buyerId, listingOrderQueueDTO.sellerId,
            listingOrderQueueDTO.buyPrice, listingOrderQueueDTO.sellPrice
        )
    }
}